storage/src/adb/keyless.rs

```diff
@@ -165,75 +160,184 @@ impl<E: Storage + Clock + Metrics, V: Codec, H: CHasher> Keyless<E, V, H> {
         )
         .await?;
 
-        // Rewind to the last commit point if necessary.
-        let mut op_index = locations_size;
-        // The most recent commit point (if any).
-        let mut last_commit_loc = None;
-        // The location of the first append operation to follow the last known commit, and the
-        // offset it wraps.
-        let mut rewind_point = None;
-        while op_index > 0 {
-            op_index -= 1;
-            let offset = locations.read(op_index).await?;
-            let section = op_index / cfg.log_items_per_section.get();
-            let op = log.get(section, offset).await?.expect("no operation found");
-            match op {
-                Operation::Commit(_) => {
-                    last_commit_loc = Some(op_index);
-                    break;
+        // Trim any locations/mmr elements that do not have corresponding operations in log.
+        let log_items_per_section = cfg.log_items_per_section.get();
+        let mut walked_back = 0;
+        let mut section_offset = None;
+        while aligned_size - walked_back > 0 {
+            let loc = aligned_size - walked_back - 1;
+            let offset = locations.read(loc).await?;
+            let section = loc / log_items_per_section;
+            match log.get(section, offset).await {
+                Ok(Some(_)) => {
+                    section_offset = Some((section, offset));
+                    break; // operation exists in the log
                 }
-                Operation::Append(_) => {
-                    rewind_point = Some((op_index, offset));
+                Ok(None) => (),
+                Err(e) => {
+                    warn!(loc, err = e.to_string(), "error finding operation in log");
                 }
-            }
+            };
+            warn!(
+                loc,
+                offset, section, "locations is ahead of log, walking back"
+            );
+            walked_back += 1;
+        }
+
+        if walked_back > 0 {
+            warn!(
+                size = aligned_size,
+                op_count = walked_back,
+                "trimming locations & mmr elements ahead of log"
+            );
+            locations.rewind(aligned_size - walked_back).await?;
+            locations.sync().await?;
+            mmr.pop(walked_back as usize).await?;
         }
-        if let Some(last_commit_loc) = last_commit_loc {
-            if last_commit_loc != locations_size - 1 {
-                // There's at least one append operation to rewind.
+        assert_eq!(mmr.leaves(), locations.size().await?);
+
+        // At this point, the tip of the locations journal & mmr, if they exist, correspond
+        // to a valid log location represented by `section_offset`.
+
```

Comment:
nit: feels weird to have this gap here. I think we can add a new comment for the following block?